File: app/admin/service/internal/biz/department.go
package biz

import (
	"context"
	"time"

	"github.com/go-kratos/kratos/v2/log"
	"github.com/hypercoze/kratos-admin/pkg/utils/uuid"
)

type Department struct {
	ID          string    `json:"id"`          // 部门ID
	ParentID    string    `json:"parentID"`    // 父部门ID，空表示根节点
	Name        string    `json:"name"`        // 部门名称
	Code        string    `json:"code"`        // 部门code
	Type        string    `json:"type"`        // 类型（company=公司 subsidiary=子公司 department=部门 position=岗位）
	Weight      int       `json:"weight"`      // 权重
	Status      string    `json:"status"`      // 状态：enabled=启用 disabled=禁用
	LeaderName  string    `json:"leaderName"`  // 负责人姓名
	LeaderPhone string    `json:"leaderPhone"` // 联系电话
	LeaderEmail string    `json:"leaderEmail"` // 邮箱
	CreatedAt   time.Time `json:"createdAt"`   // 创建时间
	UpdatedAt   time.Time `json:"updatedAt"`   // 更新时间

	Children []*Department `json:"children"`
}

type DepartmentListResult struct {
	Name   string `form:"name" query:"name"`
	Code   string `form:"code" query:"code"`
	Status string `form:"status" query:"status"`
}

type DepartmentRepo interface {
	ListDepartment(ctx context.Context, req *DepartmentListResult) ([]*Department, error)
	CreateDepartment(ctx context.Context, o *Department) error
	UpdateDepartment(ctx context.Context, o *Department) error
	DeleteDepartment(ctx context.Context, ids []string) error
	UpdateDepartmentStatus(ctx context.Context, id, status string) error
	GetDepartment(ctx context.Context, id string) (*Department, error)
}

type DepartmentUsecase struct {
	repo DepartmentRepo
	log  *log.Helper
}

func NewDepartmentUsecase(repo DepartmentRepo, logger log.Logger) *DepartmentUsecase {
	return &DepartmentUsecase{
		repo: repo,
		log:  log.NewHelper(logger),
	}
}

// 列表权限的tree
func (uc *DepartmentUsecase) DepartmentTree(ctx context.Context, req *DepartmentListResult) ([]*Department, error) {
	list, err := uc.repo.ListDepartment(ctx, req)
	if err != nil {
		return nil, err
	}
	tree := buildDepartmentTree(list)

	return tree, nil
}

func (uc *DepartmentUsecase) GetDepartment(ctx context.Context, id string) (*Department, error) {
	return uc.repo.GetDepartment(ctx, id)
}

// 创建
func (uc *DepartmentUsecase) CreateDepartment(ctx context.Context, req *Department) (*Department, error) {
	department := req

	now := time.Now()
	department.ID = uuid.GenerateXID()
	department.CreatedAt = now
	department.UpdatedAt = now

	err := uc.repo.CreateDepartment(ctx, department)
	return department, err
}

// 更新
func (uc *DepartmentUsecase) UpdateDepartment(ctx context.Context, req *Department) (*Department, error) {
	department := req

	now := time.Now()
	department.UpdatedAt = now

	err := uc.repo.UpdateDepartment(ctx, department)
	return department, err
}

// 删除
func (uc *DepartmentUsecase) DeleteDepartment(ctx context.Context, ids []string) error {
	return uc.repo.DeleteDepartment(ctx, ids)
}

// 更新状态
func (uc *DepartmentUsecase) UpdateDepartmentStatus(ctx context.Context, id, status string) error {
	return uc.repo.UpdateDepartmentStatus(ctx, id, status)
}

File: frontend/apps/web-antd/src/api/organization/department.ts
import { requestClient } from '#/api/request';
import type { AllResult } from '#/types/pagination'

export namespace OrganizationDepartmentApi {
  /** 部门 */
  export interface Department {
    /** ID */
    id: string;
    /** 父部门ID，空表示根节点 */
    parentID: string;
    /** 名称 */
    name: string;
    /** 编码 */
    code?: string;
    /** 权重 */
    weight: number;
    /** 状态：enabled=启用 disabled=禁用 */
    status: 'enabled' | 'disabled';
    /** 负责人姓名 */
    leaderName?: string;
    /** 联系电话 */
    leaderPhone?: string;
    /** 邮箱 */
    leaderEmail?: string;
    /** 创建时间 */
    createdAt?: string;
    /** 更新时间 */
    updatedAt?: string;
    /** 子级部门 */
    children?: Department[];
  }

  export interface DepartmentTreeParams {
    name?: string;
    code?: string;
    status?: 'enabled' | 'disabled';
  }
}

/**
 * 获取部门树
 */
async function getDepartmentTreeApi(params?: OrganizationDepartmentApi.DepartmentTreeParams) {
  return requestClient.get<AllResult<OrganizationDepartmentApi.Department>>(
    '/admin/v1/department/tree',
    {
      params: params
    }
  );
}

/**
 * 获取部门
 */
async function getDepartmentApi(id:string) {
  return requestClient.get<OrganizationDepartmentApi.Department>(
    `/admin/v1/department/${id}`,
  );
}

/**
 * 创建部门
 * @param data 部门数据
 */
async function createDepartmentApi(
  data: Omit<OrganizationDepartmentApi.Department, 'id' | 'children'>,
) {
  return requestClient.post('/admin/v1/department', data, {
    showSuccessMessage: true,
  });
}

/**
 * 更新部门
 *
 * @param id 部门 ID
 * @param data 部门数据
 */
async function updateDepartmentApi(
  id: string,
  data: Omit<OrganizationDepartmentApi.Department, 'id' | 'children'>,
) {
  return requestClient.put(`/admin/v1/department/${id}`, data, {
    showSuccessMessage: true,
  });
}

/**
 * 更新部门状态
 *
 * @param id 部门 ID
 * @param status 状态
 */
async function updateDepartmentStatusApi(
  id: string,
  status: OrganizationDepartmentApi.Department['status'],
) {
  return requestClient.put(`/admin/v1/department/${id}/status`, {
    status,
  });
}

/**
 * 删除部门
 * @param ids 部门 ID 集合
 */
async function deleteDepartmentApi(ids: string[]) {
  return requestClient.delete('/admin/v1/department', {
    params: {
      ids,
    },
  });
}

export {
  getDepartmentTreeApi,
  getDepartmentApi,
  createDepartmentApi,
  updateDepartmentApi,
  updateDepartmentStatusApi,
  deleteDepartmentApi,
};

